pkg/gatewayroute/resource_manager.go

```diff
@@ -0,0 +1,357 @@
+		MeshName:           ms.Spec.AWSName,
+		MeshOwner:          ms.Spec.MeshOwner,
+		Spec:               sdkGRSpec,
+		Tags:               nil,
+		VirtualGatewayName: vg.Spec.AWSName,
+		GatewayRouteName:   gr.Spec.AWSName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.GatewayRoute, nil
+}
+
+func (m *defaultResourceManager) updateSDKGatewayRoute(ctx context.Context, sdkGR *appmeshsdk.GatewayRouteData, ms *appmesh.Mesh, vg *appmesh.VirtualGateway, gr *appmesh.GatewayRoute, vsByKey map[types.NamespacedName]*appmesh.VirtualService) (*appmeshsdk.GatewayRouteData, error) {
+	actualSDKGRSpec := sdkGR.Spec
+	desiredSDKGRSpec, err := m.buildSDKGatewayRouteSpec(ctx, gr, vsByKey)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := cmpopts.EquateEmpty()
+	if cmp.Equal(desiredSDKGRSpec, actualSDKGRSpec, opts) {
+		return sdkGR, nil
+	}
+	if !m.isSDKGatewayRouteControlledByCRDGatewayRoute(ctx, sdkGR, gr) {
+		m.log.V(2).Info("skip gatewayRoute update since it's not controlled",
+			"gatewayRoute", k8s.NamespacedName(gr),
+			"gatewayRouteARN", aws.StringValue(sdkGR.Metadata.Arn),
+		)
+		return sdkGR, nil
+	}
+
+	diff := cmp.Diff(desiredSDKGRSpec, actualSDKGRSpec, opts)
+	m.log.V(2).Info("gatewayRouteSpec changed",
+		"gatewayRoute", k8s.NamespacedName(gr),
+		"actualSDKGRSpec", actualSDKGRSpec,
+		"desiredSDKGRSpec", desiredSDKGRSpec,
+		"diff", diff,
+	)
+	resp, err := m.appMeshSDK.UpdateGatewayRouteWithContext(ctx, &appmeshsdk.UpdateGatewayRouteInput{
+		MeshName:           ms.Spec.AWSName,
+		MeshOwner:          ms.Spec.MeshOwner,
+		Spec:               desiredSDKGRSpec,
+		VirtualGatewayName: vg.Spec.AWSName,
+		GatewayRouteName:   sdkGR.GatewayRouteName,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.GatewayRoute, nil
+}
+
+func (m *defaultResourceManager) deleteSDKGatewayRoute(ctx context.Context, sdkGR *appmeshsdk.GatewayRouteData, ms *appmesh.Mesh, vg *appmesh.VirtualGateway, gr *appmesh.GatewayRoute) error {
+	if !m.isSDKGatewayRouteOwnedByCRDGatewayRoute(ctx, sdkGR, gr) {
+		m.log.V(2).Info("skip mesh gatewayRoute since its not owned",
+			"gatewayRoute", k8s.NamespacedName(gr),
+			"gatewayRouteARN", aws.StringValue(sdkGR.Metadata.Arn),
+		)
+		return nil
+	}
+
+	_, err := m.appMeshSDK.DeleteGatewayRouteWithContext(ctx, &appmeshsdk.DeleteGatewayRouteInput{
+		MeshName:           ms.Spec.AWSName,
+		MeshOwner:          ms.Spec.MeshOwner,
+		VirtualGatewayName: vg.Spec.AWSName,
+		GatewayRouteName:   sdkGR.GatewayRouteName,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) updateCRDGatewayRoute(ctx context.Context, gr *appmesh.GatewayRoute, sdkGR *appmeshsdk.GatewayRouteData) error {
+	oldGR := gr.DeepCopy()
+	needsUpdate := false
+	if aws.StringValue(gr.Status.GatewayRouteARN) != aws.StringValue(sdkGR.Metadata.Arn) {
+		gr.Status.GatewayRouteARN = sdkGR.Metadata.Arn
+		needsUpdate = true
+	}
+
+	if aws.Int64Value(gr.Status.ObservedGeneration) != gr.Generation {
```

Comment:
Should it be only when ObservedGeneration < Generation?

Reply:
Technically yes as this covers that too. Shouldn't cause any issues either way. Do you see any issues with the current approach?

Reply:
I mean, once observedGeneration becomes > generation not sure how they can ever catch up. We will end up updating the resource every time even when there is no need.